### ISSUE TITLE
Consider renaming package.json in the next major version (3.0)

### ISSUE THREAD

**Issue by glen-84** [npm/npm#8222]
A default ASP.NET 5 web application has the following JSON files:

bower.json
config.json
package.json
project.json

With this, it's easy to assume that package.json is used by the [NuGet](http://nuget.org/) package manager, which of course is not true. The name is too generic.

Bower has the right idea in naming its file bower.json, can NPM please do the same?

npm.json, or something similar.

IMHO, including the word "package" implies that the project is itself an NPM package, which is obviously not always true.

I realize that this is backwards-incompatible, but that's what major versions are for, right?

You *could* support both names, but that would mean that other tools (like JSPM), would have to do the same, which may not be ideal.

**Comment by smikes** [comment]
This is an interesting suggestion.  

Unfortunately the name `package.json` is specified in the node module loading spec, (see https://nodejs.org/dist/v0.10.38/docs/api/modules.html ), which is at the highest level of stability (5; locked), so I think this is very unlikely to ever change.  

If it did change, it would require a coordinated move by both `npm` and `node`.

**Comment by othiym23** [comment]
As @smikes implies, the `package.json` name is baked into the [CommonJS specification](http://wiki.commonjs.org/wiki/Packages/1.0) which both npm and Node share (it is, in fact, the only place where npm and Node overlap, aside from npm following the Node [module resolution algorithm](https://iojs.org/api/modules.html) for installing packages.

More to the point, `package.json` is used by far more than just npm and Node at this point, with many other pieces of tooling using it as the location of their metadata and configuration. See also [this](https://docs.npmjs.com/misc/faq#so-why-is-it-the-node-modules-folder-but-package-json-file-why-not-node-packages-or-module-json) and [this](https://docs.npmjs.com/misc/faq#node-modules-is-the-name-of-my-deity-s-arch-rival-and-a-forbidden-word-in-my-religion-can-i-configure-npm-to-use-a-different-folder) from npm's FAQ. This isn't going to change.

**Comment by glen-84** [comment]
I apologize if this is a stupid question, but is Node/NPM ever going to switch to the ES6 module syntax?

**Comment by smikes** [comment]
That's absolutely not a stupid question.

Once ES6 is finalized (it's still in a draft / release candidate stage), then I believe node will move to supporting the ES6 module syntax in parallel with the CommonJS spec. 

For example you can use a module now with the CommonJS style:
```
var f = require("foo").func;
var bar = require("bar");
```

And in the future you will probably be able to do this instead:
```
import {func as f} from "foo";
import * as bar from "bar";
```

The ES6 specification is deliberately vague on some issues of module loading and resolution, because they are outside the domain of the ES language and depend on the host environment -- browser, server or embedded system.  As node moves into supporting ES6, the module loading subsystem will have to support existing (CommonJS) modules and the node_modules directory system, and also the new ES6 modules syntax.  

At present, though, I think it's safe to say that nobody is entirely sure what that will look like when it's done.

**Comment by glen-84** [comment]
Interesting.

I thought that it may be an opportunity to make other backwards-incompatible changes (like the package.json rename or a better config format like JSON5), but if you're still going to support CommonJS syntax, then I guess this will not be possible.

**Comment by othiym23** [comment]
I would be very surprised if Node ever drops support for CommonJS packages, and since there are over 140,000 packages in the npm ecosystem right now, that makes the linchpin around which all of those packages are built pretty much immovable. Realistically speaking, the window to change the name of `package.json` for npm purposes closed some time in 2011. It would be far more pain than gain to change now.

Also, to add a tiny bit to Sam's response, there are actually two specifications that govern ES6 modules – there's the syntax and static semantics of the modules, which is part of ES-262, and there's a separate specification under the aegis of the WHATWG for the module loader. The former is done – ES2015 is on track to be ratified in a couple months – but [the latter](https://github.com/whatwg/loader) is still in flux. The decoupling between the two is designed to make it easy to work with existing dependency resolution strategies, though, so I'll be surprised if the final `System.loader` implementation for Node doesn't end up mapping over the legacy CommonJS setup in some way, much like BabelJS does today.

**Comment by glen-84** [comment]
Would it make sense to have a transition period (which could last years), where:

1. ES6 modules are supported *in addition* to CJS modules.
2. JSON5 (or similar) config format is supported *in addition* to JSON config format.
3. npm.(something) file name is supported *in addition* to package.json.

Popular packages would probably "upgrade" fairly quickly (or maintain 2 versions of their packages), and all documentation would suggest using the newer technologies, meaning that most new packages would be written this way.

My point is, it seems silly to never be able to take advantage of new innovations, and to be held back because of what currently exists. Drastic changes that cut off access to 140k packages would be insane, but I think that a transition period may be a viable option.

**Comment by Fishrock123** [comment]
> name is baked into the CommonJS specification which both npm and Node share (it is, in fact, the only place where npm and Node overlap, aside from npm following the Node module resolution algorithm for installing package

A quick note: io.js (and eventually node) doesn't strictly (have to) follow the Common JS spec anymore. CommonJS is pretty dead at this point and we're starting to need things outside of it. (Like `assert.deepStrictEqual`)

**Comment by othiym23** [comment]
@Fishrock123 However, I do not see Node or npm _ever_ moving away from using `package.json` in the module resolution algorithm because of the ridiculous amounts of pain it would cause.

@glen-84 
> With this, it's easy to assume that package.json is used by the NuGet package manager, which of course is not true. The name is too generic.

> My point is, it seems silly to never be able to take advantage of new innovations, and to be held back because of what currently exists.

These are your use cases for changing the name of `package.json`, at least one of which isn't actionable.

On the other hand:

* `package.json` is one of the most strongly-established conventions in the JavaScript world.
* The npm team has little control over what version of the software is bundled with OS distributions. It's still not uncommon for us to see versions of npm as old as 1.1.x being used, because that's what's packed with old RHEL / EPEL distributions. A massive number of users are still running `npm@1.4.28` because Node.js's packaging policy didn't allow for semver-major upgrades of dependencies for patch releases of Node. Any backwards-compatibility breaking change poses massive supportability problems, not just for the npm team, but for distro vendors and enterprise ops teams.
* Every package currently on the npm registry, including a large number of heavily-used packages that are basically finished and therefore receive no updates, depends on `package.json`.
* npm is not a monolithic piece of software. `package.json` is hardcoded into about half of the 40 or so modules that compose npm.
* Whole software ecosystems, including browserify, depend on properties in a file named `package.json`
* The Node module resolution algorithm uses `package.json` to look up the main entry point for a module, and I _highly_ doubt anyone on the Node side of things is enthusiastic about the idea of messing with that.

To me, the balance is massively lopsided in favor of leaving `package.json` as is. There is _no_ benefit and a huge amount of disruption to the JavaScript community if we (tried to) make this change, and the problem as you've stated it is relatively small. While the npm team tries to be flexible about most things, there are a few immutable facts in the npm universe, and the spelling of `package.json` is one of them. npm is not going to move away from `package.json`.

**Comment by Fishrock123** [comment]
> However, I do not see Node or npm ever moving away from using package.json in the module resolution algorithm because of the ridiculous amounts of pain it would cause.

Haha, yeah, I'm not suggesting we ever would. :)

**Comment by glen-84** [comment]
@othiym23,

You completely ignored my point about supporting both side-by-side. This issue affects not only ASP, but any other non-JavaScript framework which may wish to use NPM to manage JavaScript packages.

Anyway, it's not something that's going to change until it starts affecting you or someone else on the Node/NPM team, which is unlikely to happen since most of you probably only work with JavaScript.

Thanks for your time.

**Comment by rlidwka** [comment]
> the package.json name is baked into the CommonJS specification which both npm and Node share

`npm` does not support CommonJS Packages 1.1 proposal. 

Ironically, the spec contains an example, which doesn't quite work in practice.

```json
   "overlay": {
       "node" : {"dependencies":[...]},
       "npm"  : {"scripts":{"install":"./npm-install.sh"}
     }
```

So linking to CommonJS spec is irrelevant here.

In fact, an existence of CommonJS standard is an argument for npm to change the name of the file. So it wouldn't mess with programs that **do** support CommonJS, but have an `engines` array instead of an object for example (and that's from the spec too).

**Comment by othiym23** [comment]
> You completely ignored my point about supporting both side-by-side.

I did not. The second point I enumerated in my list of concerns concerns exactly that: even if npm supports a new file with a new name alongside `package.json`, no version of npm older than the first version with support for the new name will be able to handle that package, unless the registry (or publish process) automatically transforms the package at publish time to also have a `package.json`, which (to me) renders the new filename superfluous.

This is why I mentioned that the npm registry is still regularly getting requests from versions of npm that were published three years ago: it doesn't matter how long we make the transitional period, this change will cause pain for people regardless of how long it lasts. Many people don't know that there's a difference between Node and npm, and even more don't understand that even new versions of Node 0.10 ship with an old (buggy) version of npm. Getting people to upgrade is not something that's under the npm team's control, except in an indirect way.

> Anyway, it's not something that's going to change until it starts affecting you or someone else on the Node/NPM team, which is unlikely to happen since most of you probably only work with JavaScript.

If that's what you took away from this discussion, then I did a poor job of expressing myself. My point was not that your concern is invalid, it's that it's primarily an ergonomic one, rather than something that blocks npm or NuGet use. Ergonomics are extremely important – these are tools that people have to somehow figure out, and then, frequently, use on a daily basis – so I'm not dismissing them. I was, however, stacking them against the consequences of the change, and pointing out how skewed the result is. When you've got as many users as npm has, even simple changes have unpredictable (and large) consequences. What you're proposing isn't a simple change.

@rlidwka Perhaps I should have specified that it was CJS 1.0, but the stress on CommonJS is misplaced in any case. `package.json` is the one touch point between Node and npm, and the piece of Node that it touches is the electrified third rail of Node that _almost nobody_ who's had to deal with it (including me) wants to see change again anytime soon.

**Comment by glen-84** [comment]
@othiym23,

I've just realized that we've been looking at this the wrong way. The desired change can be done in a completely backwards-compatible manner, and in fact make the system more logical.

It's simple: You use package.json for packages, and npm.json for "dependent applications" that themselves do not need to be a package. Makes sense, right?

This way, all existing and future packages will continue to work as they do now.

"npm install" will look for npm.json first, and fall back to package.json for existing "dependent applications". Dependency resolution will *only* look for package.json.

So it's one name for all packages (package.json), and one name (with a fallback) for all "dependent applications" (npm.json).

This change would not even require a major version bump.

The only issue that I can think of is other systems that rely on package.json *in dependent applications*, such as JSPM. They would require the user to keep using a package.json file until they were updated to look for both files.

If I'm forgetting something obvious, or you still don't like the idea, then I will not say anything more.
